src/crypto/tls/delegated_credentials.go

```diff
@@ -0,0 +1,482 @@
+	serPub, err := cred.marshalPublicKeyInfo()
+	if err != nil {
+		return nil, err
+	}
+	// Assert that the public key encoding is no longer than 2^24  bytes.
+	if len(serPub) > dcMaxPubLen {
+		return nil, errors.New("tls: public key is not valid")
+	}
+
+	var serLen [2]byte
+	binary.BigEndian.PutUint16(serLen[:], uint16(len(serPub)))
+	ser = append(ser, serLen[:]...)
+	ser = append(ser, serPub...)
+
+	return ser, nil
+}
+
+// unmarshalCredential decodes serialized bytes and returns a credential, if possible.
+func unmarshalCredential(ser []byte) (*Credential, error) {
+	if len(ser) < 8 {
+		return nil, errors.New("tls: delegated credential is not valid")
+	}
+
+	validTime := time.Duration(binary.BigEndian.Uint32(ser)) * time.Second
+	pubAlgo := SignatureScheme(binary.BigEndian.Uint16(ser[4:6]))
+	pubLen := binary.BigEndian.Uint16(ser[6:8])
+
+	pubKey, err := x509.ParsePKIXPublicKey(ser[8:])
+	if err != nil {
+		return nil, err
+	}
+
+	if len(ser[8:]) != int(pubLen) {
+		return nil, errors.New("tls: delegated credential is not valid")
+	}
+
+	return &Credential{validTime, pubAlgo, pubKey}, nil
+}
+
+// getCredentialLen returns the number of bytes comprising the serialized
+// credential struct inside the Delegated Credential.
+func getCredentialLen(ser []byte) (int, error) {
+	if len(ser) < 8 {
+		return 0, errors.New("tls: delegated credential is not valid")
+	}
+
+	// The validity time.
+	ser = ser[4:]
+
+	// The expCertVerfAlgo.
+	ser = ser[2:]
+
+	// The lenght of the Public Key.
+	pubLen := int(binary.BigEndian.Uint16(ser))
+	ser = ser[2:]
+
+	if len(ser) < pubLen {
+		return 0, errors.New("tls: delegated credential is not valid")
+	}
+
+	return 8 + pubLen, nil
+}
+
+// getHash maps the SignatureScheme to its corresponding hash function.
+// TODO: replace this with typeAndHashFromSignatureScheme
+func getHash(scheme SignatureScheme) crypto.Hash {
+	switch scheme {
+	case ECDSAWithP256AndSHA256:
+		return crypto.SHA256
+	case ECDSAWithP384AndSHA384:
+		return crypto.SHA384
+	case ECDSAWithP521AndSHA512:
+		return crypto.SHA512
+	default:
+		return 0 // Unknown hash function
+	}
+}
+
+// getCurve maps the SignatureScheme to its corresponding elliptic.Curve.
+// TODO: replace this with typeAndHashFromSignatureScheme
+func getCurve(scheme SignatureScheme) elliptic.Curve {
+	switch scheme {
+	case ECDSAWithP256AndSHA256:
+		return elliptic.P256()
+	case ECDSAWithP384AndSHA384:
+		return elliptic.P384()
+	case ECDSAWithP521AndSHA512:
+		return elliptic.P521()
+	default:
+		return nil
+	}
+}
+
+// prepareDelegation returns the message that the delegator is going to sign.
+// The inputs are the credential ('cred'), the DER-encoded end-entity
+// certificate ('dCert'), the signature scheme of the delegator
+// ('algo').
+func prepareDelegation(cred *Credential, dCert []byte, algo SignatureScheme, peer string) ([]byte, error) {
+	values := make([]byte, 64, 128)
+	for i := range values {
+		values[i] = 0x20
+	}
+
+	if peer == "server" {
```

Comment:
peer could be a named constant accepting only server and client.

Reply:
Added.